replica/transition.go

```diff
@@ -34,24 +70,154 @@ type Proposed struct {
 	block.Block
 }
 
-// IsTransition implements the `Transition` interface for the `Proposed` event.
+// IsTransition implements the `Transition` interface for the
+// `Proposed` event.
 func (proposed Proposed) IsTransition() {
 }
 
-// A PreVoted block has been signed and broadcast by another `Replica`.
+// A PreVoted block has been signed and broadcast by another
+// `Replica`.
 type PreVoted struct {
 	block.SignedPreVote
 }
 
-// IsTransition implements the `Transition` interface for the `PreVoted` event.
+// IsTransition implements the `Transition` interface for the
+// `PreVoted` event.
 func (preVoted PreVoted) IsTransition() {
 }
 
-// A PreCommitted polka has been signed and broadcast by another `Replica`.
+// A PreCommitted polka has been signed and broadcast by another
+// `Replica`.
 type PreCommitted struct {
 	block.SignedPreCommit
 }
 
-// IsTransition implements the `Transition` interface for the `PreCommitted` event.
+//TODO: should we have Commited as well?
+
+// IsTransition implements the `Transition` interface for the
+// `PreCommitted` event.
 func (preCommitted PreCommitted) IsTransition() {
 }
+
+func (buffer *transitionBuffer) Enqueue(transition Transition) {
+	switch transition := transition.(type) {
+	case Proposed:
+		buffer.initMapKey(transition.Height)
+		queue := buffer.buf[transition.Height]
+		if tran, ok := queue.peek(); ok {
+			switch tran.(type) {
+			case Proposed:
+			case PreVoted:
+				panic("You Enqueued a PreVoted before a Proposed")
+			case PreCommitted:
+				panic("You Enqueued a PreCommitted before a Proposed")
```

Comment:
We should not panic in these situations. The asynchronous (and potentially malicious) nature of the network means that we might _never_ receive a `Proposed` transition at a height, but we might receive a `PreVoted`. There is nothing wrong with enqueueing the `PreVoted` transition at a height, before receiving a `Proposed` at that height. The transition buffer should be able to handle this case (and ignore the enqueue if a `Proposed` is received after a `PreVoted` at a height).

Note: I am using `Proposed` and `PreVoted` as example but it holds for most transition combinations.

Reply:
Ah, so instead of panic, just ignore them.

Reply:
Note: the code is fine with getting a `PreVoted` before any `Proposed`. I just assumed something else was ensuring I didn't get a `Proposed` after I got a `PreVoted`.

Reply:
Oh, I see, this type switch is on the peek of the queue. I missed that. Yes, ignore the Transition if it is a Proposed and the queue has something "more relevant" in it. Same goes for the other Transitions.